internal/webhooks/clusterclass.go

```diff
@@ -168,6 +176,216 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
+			if alteredVar, ok := varsDiff.altered[clusterVar.Name]; ok {
+				if errs := variables.ValidateClusterVariable(clusterVar.DeepCopy(), alteredVar, field.NewPath("")); len(errs) > 0 {
+					errorInfo.altered.add(alteredVar.Name, cluster.Name)
+				}
+			}
+
+			// 2) Error if a variable removed in the update is still in use in some Clusters.
+			if _, ok := varsDiff.removed[clusterVar.Name]; ok {
+				errorInfo.removed.add(clusterVar.Name, cluster.Name)
+			}
+
+			// 3) Error if a variable has been added in the update check is no longer valid on the Cluster.
+			// NOTE: This can't occur in normal circumstances as a variable must be defined in a ClusterClass in order to be introduced in
+			// a Cluster. This check may catch errors in cases involving broken Clusters.
+			if addedVars, ok := varsDiff.added[clusterVar.Name]; ok {
+				if errs := variables.ValidateClusterVariable(clusterVar.DeepCopy(), addedVars, field.NewPath("")); len(errs) > 0 {
+					errorInfo.added.add(clusterVar.Name, cluster.Name)
+				}
+			}
+		}
+	}
+
+	// 4) Error if a required variable is not defined in every cluster using the ClusterClass.
+	for varName := range varsDiff.required {
+		clustersMissingVariable := clustersWithoutVar(clusterList, tracker.get(varName))
+		if len(clustersMissingVariable) > 0 {
+			errorInfo.required.add(varName, clustersMissingVariable...)
+		}
+	}
+	// Aggregate the errors from the validation checks and return one error message of each type for each variable with a list of violating clusters.
+
+	return aggregateErrors(errorInfo, clusterClassVariablesIndex, path)
+}
+
+// errorAggregator used to aggregate information about errors during variable comparison.
+// TODO(killianmuldoon): Make the errorAggregator generic so it can be used in cases where a single error aggregates errors related to many objects.
+type errorAggregator struct {
+	altered  variableClusterReferences
+	added    variableClusterReferences
+	removed  variableClusterReferences
+	required variableClusterReferences
+}
+
+func aggregateErrors(errs errorAggregator, clusterClassVariablesIndex map[string]int, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// Look through the errors and append aggregated error messages naming the Clusters blocking changes to existing variables.
+	if len(errs.removed) > 0 {
+		for variable, clusters := range errs.removed {
+			allErrs = append(allErrs,
+				field.Forbidden(path.Index(clusterClassVariablesIndex[variable]),
+					fmt.Sprintf("variable %s can not be removed. It is used in Clusters %v", variable, clusters),
+				))
+		}
+	}
+	if len(errs.altered) > 0 {
+		for variable, clusters := range errs.altered {
+			allErrs = append(allErrs,
+				field.Forbidden(path.Index(clusterClassVariablesIndex[variable]),
+					fmt.Sprintf("variable schema for %s can not be altered. It failed validation in Clusters %v", variable, clusters),
```

Comment:
```suggestion
					fmt.Sprintf("variable schema change for variable %s is invalid. It failed validation in Clusters %v", variable, clusters),
```
nit. Just to avoid confused users which might now think that the schema cannot be altered (as it can be altered just not to *that* schema)